control-plane-operator/controllers/hostedcontrolplane/oauth/idp_convert.go

```diff
@@ -0,0 +1,570 @@
+	defer cancel()
+
+	req, err := http.NewRequest("POST", tokenURL, body)
+	if err != nil {
+		return false, err
+	}
+	req = req.WithContext(reqCtx)
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	// explicitly set Accept to 'application/json' as that's the expected deserializable output
+	req.Header.Set("Accept", "application/json")
+
+	client := &http.Client{Transport: transport}
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	respJSON := json.NewDecoder(resp.Body)
+	respMap := map[string]interface{}{}
+	if err = respJSON.Decode(&respMap); err != nil {
+		// only log the error, some OIDCs ignore/don't implement the Accept header
+		// and respond with HTML in case they don't support password credential grants at all
+		klog.Errorf("failed to JSON-decode the response from the OIDC server's token endpoint (%s): %v", tokenURL, err)
+		oidcPasswordCheckCache.Set(secret.ResourceVersion, false, oidcPasswordTTL)
+		return false, nil
+	}
+
+	if errVal, ok := respMap["error"]; ok {
+		oidcPasswordCheckCache.Set(secret.ResourceVersion, errVal == "invalid_grant", oidcPasswordTTL) // wrong password, but password grants allowed
+	} else {
+		_, ok = respMap["access_token"] // in case we managed to hit the correct user
+		oidcPasswordCheckCache.Set(secret.ResourceVersion, ok, oidcPasswordTTL)
+	}
+
+	result, _ := oidcPasswordCheckCache.Get(secret.ResourceVersion)
+	return result.(bool), nil
+}
+
+type openIDProviderJSON struct {
+	AuthURL     string `json:"authorization_endpoint"`
+	TokenURL    string `json:"token_endpoint"`
+	UserInfoURL string `json:"userinfo_endpoint"`
+}
+
+func isValidURL(rawurl string, optional bool) bool {
+	if len(rawurl) == 0 {
+		return optional
+	}
+
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return false
+	}
+
+	return u.Scheme == "https" && len(u.Host) > 0 && len(u.Fragment) == 0
+}
+
+func createFileStringSource(filepath string) configv1.StringSource {
+	return configv1.StringSource{
+		StringSourceSpec: configv1.StringSourceSpec{
+			File: filepath,
+		},
+	}
+}
+
+func osinv1EncodeOrDie(obj runtime.Object) []byte {
+	bytes, err := runtime.Encode(osinv1Encoder, obj)
+	if err != nil {
+		panic(err) // indicates static generated code is broken, unrecoverable
+	}
+	return bytes
+}
+
+func transportForCARef(kclient crclient.Client, namespace, name, key string) (http.RoundTripper, error) {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	if err := kclient.Get(context.Background(), crclient.ObjectKeyFromObject(cm), cm); err != nil {
+		return nil, err
+	}
+	caData := []byte(cm.Data[key])
+	if len(caData) == 0 {
+		caData = cm.BinaryData[key]
+	}
+	if len(caData) == 0 {
+		return nil, fmt.Errorf("config map %s/%s has no ca data at key %s", namespace, name, key)
+	}
+
+	// copy default transport
+	transport := net.SetTransportDefaults(&http.Transport{
```

Comment:
Need to set transport and dialer timeout values

Reply:
So timeouts are already configured for the transport by the call to k8s net.SetTransportDefaults. I am using my own timeout in the context for the request.

Reply:
Makes sense — now that the stdlib client is context aware that probably removes the need to deal with fine grained timeout values except in unusual circumstances